crypto/hashtable/hashtable.c

```diff
@@ -736,19 +805,32 @@ int ossl_ht_delete(HT *h, HT_KEY *key)
             continue;
         if (compare_hash(hash, h->md->neighborhoods[neigh_idx].entries[j].hash)
             && match_key(key, &v->value.key)) {
-            if (!CRYPTO_atomic_store(&h->md->neighborhoods[neigh_idx].entries[j].hash,
-                                     0, h->atomic_lock))
-                break;
+            if (ossl_ht_get_cfg_no_rcu(h)) {
+                if (!CRYPTO_atomic_store(&h->md->neighborhoods[neigh_idx].entries[j].hash,
+                                         0, h->atomic_lock))
+                    break;
+                ossl_rcu_assign_ptr(&h->md->neighborhoods[neigh_idx].entries[j].value, &nv);
+            } else {
+                h->md->neighborhoods[neigh_idx].entries[j].hash = 0;
+                h->md->neighborhoods[neigh_idx].entries[j].value = NULL;
+            }
             h->wpd.value_count--;
-            ossl_rcu_assign_ptr(&h->md->neighborhoods[neigh_idx].entries[j].value,
-                                &nv);
             rc = 1;
             break;
         }
     }
     if (rc == 1) {
-        ossl_rcu_call(h->lock, free_old_entry, v);
+        if (!ossl_ht_get_cfg_no_rcu(h))
+            ossl_rcu_call(h->lock, free_old_entry, v);
+        else
+            free_old_entry(v);
         h->wpd.need_sync = 1;
     }
+
     return rc;
 }
+
+int ossl_ht_get_cfg_no_rcu(HT *htable)
+{
+    return htable->config.no_rcu;
+}
```

Comment:
Please make this a static function. I do not see why this should be exported outside of this .c file and furthermore it allows the function to be properly inlined. (And remove the ossl_ prefix when making it static.)

Reply:
Hmm... Then mark the function static ossl_inline and put the implementation into the header.